### ISSUE TITLE
Scale to zero

### ISSUE THREAD

**Issue by simonw** [aws/apprunner-roadmap#9]
I have a serious side-project habit - I often have dozens of side projects on the go at once.

As such, I really appreciate scale-to-zero services like Google Cloud Run and Vercel, where if my project isn't getting any traffic at all it costs me nothing (or just a few cents a month in storage costs) - then it spins up a server when a request comes in, with a cold-start delay of a few seconds before it starts serving traffic.

I would love it if App Runner could do this! It looks like at the moment you have to pay for a minimum of one running instance.

**Comment by danthegoodman1** [comment]
Scale to zero is really important for small projects that don’t need 24/7 compute running, and especially contractor work. Besides that microservices that don’t need to be running all the time, and side projects where someone wants a full container and is willing to deal with cold starts (like lambda but not being constrained to API gateway or lambda utilities in the container).

scale to 0 is the only thing that prevents me from using it.

**Comment by timanderson** [comment]
It scales to just $0.007 per GB/hour if the application is idle as far as I can tell, no vCPU cost. Or there is a PauseService API call to eliminate that cost too. If you had a batch job, you could call ResumeService at the start and PauseService at the end?

**Comment by danthegoodman1** [comment]
Yeah but that’s still $5/month more than cloud run, and if I’m using such a heavily managed service I wouldn’t want to automate pausing and resuming myself (with batch it would be fine but for an api it would be very difficult)

**Comment by Munawwar** [comment]
What should happen when a request is made when instances are zero? Currently (if service is paused) the root url gives a http status code 404. Do you want the end-point to wait a bit from responding for some time to give a chance to spawn an instance and respond?

**Comment by danthegoodman1** [comment]
> What should happen when a request is made when instances are zero? Currently (if service is paused) the root url gives a http status code 404. Do you want the end-point to wait a bit from responding for some time to give a chance to spawn an instance and respond?

Yeah, basically a cold start similar to lambda. That’s how cloud run does it.

**Comment by danthegoodman1** [comment]
I’ll also add it would be really good for many users to not be forced to scale to zero like cloud run does. Like if we had some field to set “minimum containers” or something similar to lambda provisioned concurrency, because there are some use cases where you never want a cold start.

**Comment by mwarkentin** [comment]
@danthegoodman1 you can configure the minimum "provisioned" containers which stay active (paying for mem only, not CPU) - except you can only set that to `>= 1`.

It would be nice if you could leave that at `1` if you wanted to remove cold start, or `0` if you wanted to optimize for costs and were ok with some latency when the first request comes in to the system after its scaled down.

**Comment by nelsonjchen** [comment]
@danthegoodman1 You aren't forced to scale to zero in Cloud Run and can set minimum instances that even charge/run at the "idle rate":

https://cloud.google.com/run/docs/configuring/min-instances

**Comment by danthegoodman1** [comment]
> @danthegoodman1 You aren't forced to scale to zero in Cloud Run and can set minimum instances that even charge/run at the "idle rate":
> 
> https://cloud.google.com/run/docs/configuring/min-instances

This only used to be available if you were using cloud run for anthos, didn’t realize they updated it, thanks.

**Comment by danthegoodman1** [comment]
> @danthegoodman1 you can configure the minimum "provisioned" containers which stay active (paying for mem only, not CPU) - except you can only set that to `>= 1`.
> 
> It would be nice if you could leave that at `1` if you wanted to remove cold start, or `0` if you wanted to optimize for costs and were ok with some latency when the first request comes in to the system after its scaled down.

Yep just wanted to make sure we still kept that feature just in case!

**Comment by flibustenet** [comment]
I would use scaling to zero for dev stack. Also if i want to show the new version of my app to my customer, he can then look at the app just when he want.

**Comment by 486** [comment]
Apart from smaller projects, scale to zero would be super useful for development workflows. Imagine many developers deploying code branches for testing. Right now, they would have to consciously deprovision the service when they are not working on it. 

With scale to zero, there would be no costs when they aren’t working (= not sending requests to their personal deployment). And cold start latency isn’t relevant in this scenario.

**Comment by nelsonjchen** [comment]
I'm surprised this hasn't been mentioned before but a turnoff of GCP is that they don't have an "Amazon Aurora Serverless" equivalent to go along with Cloud Run. 

Scale to Zero AppRunner + Amazon Aurora Serverless would be a dream.

**Comment by danthegoodman1** [comment]
> I'm surprised this hasn't been mentioned before but a turnoff of GCP is that they don't have an "Amazon Aurora Serverless" equivalent to go along with Cloud Run.
> 
> Scale to Zero AppRunner + Amazon Aurora Serverless would be a dream.

300 IQ right there, that’s an awesome idea. Also dynamodb would work too but aurora serverless (v2 Postgres plz) would be a real separator

**Comment by tomaszdudek7** [comment]
Seconding that 300IQ idea. And that scale-to-zero-with-minimal-cold-start idea.

**Comment by tomaszdudek7** [comment]
Seconding that 300 IQ statement. We need that! GCP Cloud Run is ahead. Don't make us use their product.

**Comment by danthegoodman1** [comment]
This is also amazing for Slack bots btw. I'd move our team Slack bot here in an instant if we could scale to 0.

**Comment by stephanoparaskeva** [comment]
@nelsonjchen @danthegoodman1 Why is it not yet possible to use App Runner + Aurora PostgreSQL Serverless?

I am trying to deploy an API on app on App Runner, and the API is supposed to connect to the Aurora PostgreSQL Serverless endpoint, but I cant get it to connect (locally or on App Runner). Does this mean it doesnt work?

**Comment by danthegoodman1** [comment]
@stephanoparaskeva make sure you’re in the same vpc, have proper security groups and routing tables, or in your case it sounds like enabled public access if you’re trying to test locally (although please don’t do in production)

**Comment by ashkue** [comment]
@stephanoparaskeva Aurora Serverless can only work within VPC with a private IP. AppRunner does not support VPC integration yet, however it is on the roadmap: https://github.com/aws/apprunner-roadmap/issues/1

**Comment by nelsonjchen** [comment]
> @nelsonjchen @danthegoodman1 Why is it not yet possible to use App Runner + Aurora PostgreSQL Serverless?
> 
> 
> 
> I am trying to deploy an API on app on App Runner, and the API is supposed to connect to the Aurora PostgreSQL Serverless endpoint, but I cant get it to connect (locally or on App Runner). Does this mean it doesnt work?

I don't think I said it wasn't possible. You likely have connectivity issues not related to this lack of scale to zero issue like @danthegoodman1 said.

**Comment by stephanoparaskeva** [comment]
> @stephanoparaskeva Aurora Serverless can only work within VPC with a private IP. AppRunner does not support VPC integration yet, however it is on the roadmap: #1

Ah ok so I should use a public DB for the time being.

- Once VPC is released -- if both App Runner and Aurora are in the same VPC, should it just connect via endpoint + user + password?

- Also how does one connect to Aurora using a locally running version of their API (is this possible)?

Thanks for the swift response!

**Comment by nelsonjchen** [comment]
> > @stephanoparaskeva Aurora Serverless can only work within VPC with a private IP. AppRunner does not support VPC integration yet, however it is on the roadmap: #1
> 
> 
> 
> Ah ok so I should use a public DB for the time being.
> 
> 
> 
> - Once VPC is released -- if both App Runner and Aurora are in the same VPC, should it just connect via endpoint + user + password?
> 
> 
> 
> - Also how does one connect to Aurora using a locally running version of their API (is this possible)?
> 
> 
> 
> Thanks for the swift response!

Could you take this conversation elsewhere? This issue is about scale to zero.

**Comment by toricls** [comment]
Just for clarification, but you can use Aurora Serverless v1 without VPC by using its Data API :)

https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html

**Comment by JonMarbach** [comment]
I think there's some relevant discussion in https://github.com/aws/containers-roadmap/issues/1017 on scale-to-zero for Fargate which is probably applicable here too. I made a more detailed post RE Fargate, but summing up quickly: a few seconds (or more maybe) of cold-start latency would be OK for me, and I can send a ping to the service to mitigate that prior to hitting it full-force.

**Comment by CarlosDomingues** [comment]
Just adding my 2 cents. Lots of backoffice-like and data applications could not care less about latency. Actually being able to scale to zero would be amazing for my use cases. This could be a configurable feature and the UI could say explicitly how scaling memory to zero affects latency.

**Comment by iBobik** [comment]
I use currently Heroku for this use-case - it scales to zero on free accounts.

I have not tried it yet, but because Heroku runs on AWS it could have low latency to Aurora Serverless databases.

**Comment by isae** [comment]
I think that "what about the cold start" argument is unrelated to the topic, because you'll expect cold starts in any service providing autoscaling, that's just the way programs work (especially ones running in a virtual machine).

**Comment by phishy** [comment]
If AppRunner doesn't scale to 0, why does mine show 0 instances when there is no traffic?

**Comment by jvisker** [comment]
@phishy AppRunner can scale _active_ instances to zero. However, you can have at a minimum 1 _inactive_ instance. You are charged for the memory of the inactive instances.

**Comment by deanhiller** [comment]
Coming from GCP Cloud run, and production min instance=1 while in preproduction and staging minInstance 0, this was critical as our company grew and grew with huge cost savings since staging/preprod was never used every night so it basically auto-shutdown things.   How is this still not on parity with cloud run?  (We are actively thinking of switching to GCP right now - not to mention it took us forever to find App Runner to begin with being referred to AWS fargate, then Beanstalk and I finally stumbled on App Runner by a forum post I made).  

Is this going to be done(scale to 0) so we are not billed anything while in a state of off?

ps. We do not want 404's.  we expect first request slowness as instance boots up and then it works fast and we are ok with that in staging/preproduction where we match production for # of services.

**Comment by ColeDCrawford** [comment]
Would be great to have this. I have a lot of apps which could benefit from scale-to-zero. Internal things like an RDS bastion that I would occasionally use and otherwise need to run a t3.nano or similar.

**Comment by fracampit** [comment]
Worth mentioning here:
I have seen other people complaining about the fact that if you App Runner hosted app stays idle for some time (hours), the instances scale down to zero even if you have set the minimum scaling to 1.
I am currently in this situation and I am unable to bring the service back up. Just get a 502 when I try to access it. It has also stopped writing to Cloudwatch.

**Comment by iBobik** [comment]
Would you use proxy service what can start your server like this? Then it could stop server after X minutes of inactivity.

https://user-images.githubusercontent.com/614232/194722900-e944cbb3-f4a7-4c7d-b72e-c91ef81119c0.mov

If yes, [write me an e-mail](mailto:honza@poboril.cz) about your use-case.

**Comment by andrew-belac** [comment]
> > What should happen when a request is made when instances are zero? Currently (if service is paused) the root url gives a http status code 404. Do you want the end-point to wait a bit from responding for some time to give a chance to spawn an instance and respond?
> 
> Yeah, basically a cold start similar to lambda. That’s how cloud run does it.

CloudRun does just this.  The API is a Request Event which triggers starting the container instance if needed on demand.  My wifes website runs like this and hardly anyone ever visits it yet you would never notice any delay.  If an instance is up and not too busy then it just routes to the available instance.  You can set instance scale limits.  But this way you pay for computer actually being used without haing to pay per invocation.  

This uses Knative from from Kubernetes standard.  Serverless HTTP/Events/Build Jobs.  An amazing API.

**Comment by forresthopkinsa** [comment]
KNative +1000

**Comment by algoflows** [comment]
Over a year later and still no official resolution from AWS, moving back to GCP. I thought AWS were all about the customer, yet nothing has changed when customers are asking for specific changes.

**Comment by nelsonjchen** [comment]
I'm unsubscribing from this issue since I don't really run anything on AWS at the moment professionally anymore. That said, I do want to leave a funny anecdote about one competition. 

fly.io doesn't have scale to zero containers. They still have the same problem as apprunner. They do have scale to zero ec2-like-machines though! You can have a scale to zero Minecraft server. Via AWS's own firecracker too! Opposite world! How bizarre! 

They ain't no big cloud, but it's just funny to me. Hopefully AWS might see to it they can catch up to organizations using their own products to make scale to zero products.

**Comment by algoflows** [comment]
Surprised to see no progress and no response. I've had better responses from core GCP Cloud team members on youtube threads.

**Comment by iomarcovalente** [comment]
Subscribing, I also would be very interested in this to happen

**Comment by jdrphillips** [comment]
I really would like any scale to zero container service, that isn't lambda. I am willing to pay for from-cold delays but not repeated lamda delays.

**Comment by suzukieng** [comment]
Would really like scale to zero. I have AppRunner instances for multiple environment (dev/staging/prod). Dev and Staging are usually only needed when I'm actively developing or testing the next release. It would be nice not to be charged for these "idle" instances.

**Comment by ebg1223** [comment]
This would be a big deal coming from GCP Cloud Run!

**Comment by algoflows** [comment]
And another year goes by and not a single reply from the AWS team... insane.

**Comment by tornikeo** [comment]
Ah $H!%. It's never a good sign to find to an open issue with 300+ upvotes. Guess I gotta switch my approach :laughing:

**Comment by benkehoe** [comment]
@alexanderwink can you provide a link to that page in the docs?

**Comment by alexanderwink** [comment]
Says so right on the pricing page https://aws.amazon.com/apprunner/pricing/

**Comment by gabrielboucher** [comment]
> Just to let everyone know. When the app doesn't receive any new requests for a while the CPU will throttle down to close to zero. At that time you will only pay for the provisioned memory and not for CPU. 

This is still ~5$/month per instance with 1GB of memory. Not that it's much but it adds up quickly, especially for apps that are used once a month or less. 

My real need here is a cold standby mode.

**Comment by matteocontrini** [comment]
> This is still ~5$/month per instance with 1GB of memory.

Plus, in eu-central-1 the price is higher and becomes ~$6,4/GB. If you want 1 vCPU you're forced to reserve 2 GB RAM which means $13/month just for the idle app.

**Comment by iBobik** [comment]
> Would you use service what can start your server like this? Or it could not show a button, but start on every request. Then it could stop server after X minutes of inactivity. Could start/stop instance, app runner, container, ...
> 
>  [start.page.moqup.mov ](https://user-images.githubusercontent.com/614232/194722900-e944cbb3-f4a7-4c7d-b72e-c91ef81119c0.mov)
> 
> If yes, [write me an e-mail](mailto:honza@poboril.cz) about your use-case. I consider creating it, but I does not want to work on it only for me. :-)

**Comment by CarlosDomingues** [comment]
Adding to what @gabrielboucher said:

Summing prod and staging environments, my company runs 250+ Streamlit dashboards, APIs and data pipelines that are infrequently acessed (anything from a few times a day to one time monthly).

Assuming our average container has 2vCPUs + 2GB of memory (and my maths are correct), we are talking about ~ 2500 USD monthly just for provisioned workloads. Actual bill would be higher as each container would cost ~ 0.14 per hour.

Nowadays we use Kubernetes + Knative + Karpenter with EC2 Spot Instances and the equivalent part of our AWS bill is significantly cheaper.

**Comment by mreferre** [comment]
This isn't meant to be the solution for what's being discussed in the last 10-ish updates in this thread but if your application usage is so sparse (e.g. one hit a month) have you considered using Lambda with the [Web adapter](https://github.com/awslabs/aws-lambda-web-adapter)? This [blog post](https://aws.amazon.com/blogs/compute/previewing-environments-using-containerized-aws-lambda-functions/) walks you through the why and the how. 

The TL/DR is that you would just need to add this entry to your Dockerfile to run your Python based application unmodified in Lambda:
```
COPY --from=public.ecr.aws/awsguru/aws-lambda-adapter:0.7.0 /lambda-adapter /opt/extensions/lambda-adapter
```
I have used it [here](https://github.com/mreferre/spin-wasm-multi-compute) to demonstrate how to run the same unmodified container image in Fargate and Lambda. 

If this works for you basically have a scale to zero service that, at that level of usage, is going to be free given the generous Lambda free-tier (assuming you are not using Lambda already and you are ok with experiencing a cold start). 

Again I am not suggesting this is the solution for the requirement you are raising, but perhaps some of your use cases may be served using this approach.

**Comment by mreferre** [comment]
> Adding to what @gabrielboucher said:
> 
> Summing prod and staging environments, my company runs 250+ Streamlit dashboards, APIs and data pipelines that are infrequently acessed (anything from a few times a day to one time monthly).
> 
> Assuming our average container has 2vCPUs + 2GB of memory (and my math is correct), we are talking about ~ 2500 USD monthly just for provisioned workloads. Actual bill would be much higher as each running container would cost ~ 0.14 per hour.

For the records, my understanding is that Streamlit leverages websockets connections App Runner does not support it (yet). 

See [here](https://github.com/aws/apprunner-roadmap/issues/13) and [here](https://discuss.streamlit.io/t/new-deployment-option-aws-app-runner/13084).

**Comment by CarlosDomingues** [comment]
@mreferre Thanks for the links, that's some really cool stuff!

We've tried Lambdas in the past but the timeout limits + developer experience were show stoppers for us. Might reconsider for the subset of our APIs that don't require long running jobs.

**Comment by anandhu-renie** [comment]
I'm kinda confused with the pricing of App Runner. Idle instances are when I'm not getting any traffic. But when does it become idle?

**Comment by alexanderwink** [comment]
> I'm kinda confused with the pricing of App Runner. Idle instances are when I'm not getting any traffic. But when does it become idle?

When there is no incoming traffic to the application there is ramp-down windows for aboutn 60 seconds. After that you will see that the active instances metric goes down to 0. At this point you only pay for provisioned memory.

**Comment by anandhu-renie** [comment]
Got it thanks!

**Comment by TreyWW** [comment]
400+ votes and this still isn't even being thought about.. Wow. I'd love to hear more from AWS on whether this will soon come. I'm guessing it wont because it won't directly help AWS really, they wont really gain anything from adding this

**Comment by wesmontgomery** [comment]
Can we get an update on this please? Hoping this will be moved to "Researching" soon...

@backnol-aws @snnles @jsheld @lazarben @scuw19 @amitgupta85  @akshayram-wolverine

**Comment by larryjkl** [comment]
I don't know anything, but just thinking about it and reading the comments, and watching the re:Invent videos on how they built and designed this it might be that scaling to zero just isn't feasible in App Runner? They did go to the trouble to scale down to zero CPU ( but keep the memory ). The functionality they do provide that is vaguely similar ( pause / unpause ) takes about 40s for it to become active after it was paused for my very simple app which seems way too long for a cold start. It's frustrating when they don't give any feedback on things, but I'll be surprised if they are able to pivot to provide it scale to true zero ( although it would be great ).

**Comment by masterbater** [comment]
I like this people @efekarakus @iamhopaul123 working on aws copilot, I hope this brilliant people could work on aws app runner roadmap like hipaa and scale to zero

**Comment by BreezyPod** [comment]
I spent a couple of weeks building this App Runner replacement using EC2: [BreezyPod](https://breezypod.dev/)

It has scale to 0 for free and supports WebSockets so it works for Streamlit. Feel free to give it a spin and let me know what you all think!

**Comment by FANMixco** [comment]
This is super important to be competitive against Google Cloud Run or Azure Container Instances!

**Comment by HatmanStack** [comment]
Would  help expand the moat, will Cloud Run until completed.

**Comment by tfranssen** [comment]
This would be perfect!

**Comment by karimz1** [comment]
I think this is needed ASAP, otherwise many users might avoid adopting it.
The alternative would be setting everything up manually (for example, wiring a Lambda to trigger Fargate, scaling down to zero on certain events, etc.). Realistically, very few people will do all of that for small applications.

Since AWS is designed to automate this type of workflow, it would be great to see native support here.

If I were at AWS, I’d implement this as an internal service layer that abstracts the automation logic—so users don’t need to manage event triggers, scaling, or orchestration themselves. If they want 😀

Cheers
Karim
Cloud Solution Architect

**Comment by tomaszbk** [comment]
I'm choosing GCP precisely because of this.
